File: src/main/java/edu/practice/datastructures/Trie.java
package edu.practice.datastructures;

import java.util.HashMap;
import java.util.Map;

public class Trie {

    private static class Node {
        Map<Character, Node> children;
        boolean isEndOfWord;

        public Node() {
            children = new HashMap<>();
            isEndOfWord = false;
        }
    }

    private final Node root;
    private int size;

    public Trie() {
        root = new Node();
        size = 0;
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return (size == 0);
    }

    public void insert(String word) {
        if (word == null) { return; }

        Node currentNode = root;

        for (char c : word.toCharArray()) {
            currentNode = currentNode.children.computeIfAbsent(c,
                    character -> new Node());
        }

        if (!currentNode.isEndOfWord) {
            // Word is not inserted yet and last character is not marked as end of word
            currentNode.isEndOfWord = true;
            size++;
        }
    }

    public boolean search(String word) {
        if (word == null) { return false; }

        Node currentNode = root;

        for (char character : word.toCharArray()) {
            currentNode = currentNode.children.get(character);

            if (currentNode == null) {
                return false;
            }
        }

        return currentNode.isEndOfWord;
    }

    public boolean startsWith(String prefix) {
        if (prefix == null) { return false; }

        Node currentNode = root;

        for (char character : prefix.toCharArray()) {
            currentNode = currentNode.children.get(character);

            if (currentNode == null) {
                return false;
            }
        }

        return true;
    }

    public boolean delete(String word) {
        if (word == null) { return false; }
        return deleteWordFrom(root, word, 0);
    }

    private boolean deleteWordFrom(Node currentNode, String word, int index) {
        if (index == word.length()) {
            // Word is matched
            if (!currentNode.isEndOfWord) {
                return false;
            }

            currentNode.isEndOfWord = false;
            size--;
            return currentNode.children.isEmpty();
        }

        char character = word.charAt(index);
        Node nextNode = currentNode.children.get(character);

        if (nextNode == null) {
            return false;
        }

        boolean shouldDeleteCurrentNode = deleteWordFrom(nextNode, word, index + 1);

        if (shouldDeleteCurrentNode) {
            currentNode.children.remove(character);
            return currentNode.children.isEmpty() && !currentNode.isEndOfWord;
        }

        return index == 0;
    }
}

File: src/main/kotlin/datastructures/Trie.kt
package edu.practice.datastructures

class Trie {

    private data class Node(
        val children: MutableMap<Char, Node> = hashMapOf(),
        var isEndOfWord: Boolean = false
    )

    private val root: Node = Node()
    private var size: Int = 0

    fun insert(word: String) {
        var currentNode = root

        for (char in word) {
            currentNode = currentNode.children.computeIfAbsent(char) { Node() }
        }

        if (!currentNode.isEndOfWord) {
            currentNode.isEndOfWord = true
            size++
        }
    }

    fun search(word: String): Boolean {
        var currentNode = root

        for (char in word) {
            currentNode = currentNode.children[char] ?: return false
        }

        return currentNode.isEndOfWord
    }

    fun startsWith(prefix: String): Boolean {
        var currentNode = root

        for (char in prefix) {
            currentNode = currentNode.children[char] ?: return false
        }

        return true
    }

    fun delete(word: String): Boolean {
        return deleteFromWord(root, word)
    }

    private fun deleteFromWord(currentNode: Node, word: String, index: Int = 0): Boolean {
        if (index == word.length) { // Word matched
            if (!currentNode.isEndOfWord) { // But not end of word
                return false
            }

            currentNode.isEndOfWord = false
            size--
            return currentNode.children.isEmpty()
        }

        val char = word[index]
        val nextNode = currentNode.children[char] ?: return false

        val shouldDeleteCurrentNode = deleteFromWord(nextNode, word, index + 1)
        if (shouldDeleteCurrentNode) {
            currentNode.children.remove(char)
            return currentNode.children.isEmpty() && !currentNode.isEndOfWord
        }

        return index == 0
    }

    fun size(): Int = size

    fun isEmpty(): Boolean = (size == 0)
}